sdk_extensions/zpages/src/main/java/io/opentelemetry/sdk/extensions/zpages/TracezSpanProcessor.java

```diff
@@ -0,0 +1,193 @@
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.sdk.extensions.zpages;
+
+import com.google.common.annotations.VisibleForTesting;
+import io.opentelemetry.sdk.common.export.ConfigBuilder;
+import io.opentelemetry.sdk.trace.ReadableSpan;
+import io.opentelemetry.sdk.trace.SpanProcessor;
+import io.opentelemetry.trace.SpanId;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A {@link SpanProcessor} implementation for the traceZ zPage.
+ *
+ * <p>Configuration options for {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}
+ * can be read from system properties, environment variables, or {@link java.util.Properties}
+ * objects.
+ *
+ * <p>For system properties and {@link java.util.Properties} objects, {@link
+ * io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor} will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code otel.ssp.export.sampled}: sets whether only sampled spans should be exported.
+ * </ul>
+ *
+ * <p>For environment variables, {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}
+ * will look for the following names:
+ *
+ * <ul>
+ *   <li>{@code OTEL_SSP_EXPORT_SAMPLED}: sets whether only sampled spans should be exported.
+ * </ul>
+ */
+@ThreadSafe
+public final class TracezSpanProcessor implements SpanProcessor {
+  private final Map<SpanId, ReadableSpan> runningSpanCache;
+  private final Map<SpanId, ReadableSpan> completedSpanCache;
+  private final boolean sampled;
+
+  /**
+   * Constructor for {@link io.opentelemetry.sdk.extensions.zpages.TracezSpanProcessor}.
+   *
+   * @param sampled report only sampled spans.
+   */
+  public TracezSpanProcessor(boolean sampled) {
+    runningSpanCache = new HashMap<>();
+    completedSpanCache = new HashMap<>();
+    this.sampled = sampled;
+  }
+
+  @Override
+  public void onStart(ReadableSpan span) {
+    synchronized (this) {
+      runningSpanCache.put(span.getSpanContext().getSpanId(), span);
+    }
+  }
+
+  @Override
+  public boolean isStartRequired() {
+    return true;
+  }
+
+  @Override
+  public void onEnd(ReadableSpan span) {
+    SpanId id = span.getSpanContext().getSpanId();
+    synchronized (this) {
+      runningSpanCache.remove(id);
+      if (!sampled || span.getSpanContext().getTraceFlags().isSampled()) {
+        completedSpanCache.put(id, span);
```

Comment:
It looks like this completedSpanCache will grow without bounds. We need to make sure that we don't hold references to these spans indefinitely.

Reply:
What should the maximum size be? OpenCensus uses `Integer.MAX_VALUE.`

Reply:
It looks like 10 per bucket to me:  https://github.com/census-instrumentation/opencensus-java/blob/68e19edba8da513d20d2b62f6803463fcb02fd4f/impl_core/src/main/java/io/opencensus/implcore/trace/export/InProcessSampledSpanStoreImpl.java#L42

Where are you seeing Integer.MAX_VALUE?

Reply:
For the [running span store](https://github.com/census-instrumentation/opencensus-java/blob/a3d092a29ff44e228a0b08d878290201241e389c/contrib/zpages/src/main/java/io/opencensus/contrib/zpages/ZPageHandlers.java#L213), they use 'Integer.MAX_VALUE'.

Reply:
ah, yeah, I'm not worried about the running spans. They will be naturally limited by the SDK itself.

Reply:
Ah, I see. Then, I'll go ahead and modify the implementation to use 10 and 5 per bucket for the completed spans.

Reply:
I updated the implementation with a class similar to what OpenCensus uses. Let me know what you think!